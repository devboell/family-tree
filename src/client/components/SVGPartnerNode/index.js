import React from 'react'
import PropTypes from 'prop-types'
import SVGNode from 'components/SVGNode'
import SVGNodeInfo from 'components/SVGNodeInfo'
import SVGPartnerLink from 'components/SVGPartnerLink'

export const cumulatedSubTreeWidths = node =>
  node.children.reduce((acc, cur) =>
    [...acc, acc[acc.length - 1] + cur.totalWidth]
  , [0])

const SVGPartnerNode = ({
  node,
  y,
  offsetX,
  previousSubTreeWidth,
  dimensions,
}) => {
  const { boxWidth, boxHeight, marginX } = dimensions
  const x = offsetX + (node.totalWidth / 2) + marginX

  return (
    <g>
      <SVGNodeInfo
        {...{ x, y, boxWidth, boxHeight }}
        info={node.value}
      />
      <SVGPartnerLink
        originX={x}
        originY={y + (boxHeight / 2)}
        {...{ dimensions, previousSubTreeWidth, offsetX }}
      />
      {node.children.map((child, i) =>
        <SVGNode
          key={`node-child-${child.value}`}
          node={child}
          offsetX={offsetX + cumulatedSubTreeWidths(node)[i]}
          linkTo={{ x: x - (marginX / 2), y: y + (boxHeight / 2) }}
          {...{ dimensions }}
        />,
      )}
    </g>
  )
}
SVGPartnerNode.propTypes = {
  node: PropTypes.shape({}).isRequired,
  y: PropTypes.number.isRequired,
  offsetX: PropTypes.number.isRequired,
  previousSubTreeWidth: PropTypes.number.isRequired,
  dimensions: PropTypes.shape({}).isRequired,
}


export default SVGPartnerNode
