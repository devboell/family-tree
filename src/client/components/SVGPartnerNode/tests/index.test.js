import React from 'react'
import SVGPartnerNode from '../index'

const props = {
  node: {
    value: 'John',
    level: 2,
    children: [],
  },
  y: 100,
  offsetX: 100,
  previousSubTreeWidth: 200,
  dimensions: {
    boxWidth: 80,
    boxHeight: 80,
    marginX: 50,
  },
}

it('SVGPartnerNode, renders component', () => {
  const wrapper = shallow(<SVGPartnerNode {...props} />)
  expect(wrapper).toMatchSnapshot()
})
