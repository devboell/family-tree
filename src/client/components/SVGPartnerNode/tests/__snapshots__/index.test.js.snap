// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SVGPartnerNode, renders component 1`] = `
<g>
  <SVGNodeInfo
    boxHeight={80}
    boxWidth={80}
    info="John"
    x={NaN}
    y={100}
  />
  <SVGPartnerLink
    dimensions={
      Object {
        "boxHeight": 80,
        "boxWidth": 80,
        "marginX": 50,
      }
    }
    offsetX={100}
    originX={NaN}
    originY={140}
    previousSubTreeWidth={200}
  />
</g>
`;
